gcc-11.3.0/libcpp/mkdeps.c

```diff
@@ -725,16 +726,255 @@ calculate_sha256_omnibor_with_contents (std::string contents,
+			      unsigned hash_func_type)
+{
+  if (hash_func_type != 0 && hash_func_type != 1)
+    return false;
+
+  mkdirat (dfd, "metadata", S_IRWXU);
+
+  std::string path_metadata = res_dir + "/metadata";
+
+  DIR *dir_metadata = opendir (path_metadata.c_str ());
+  if (dir_metadata == NULL)
+    return false;
+
+  int dfd_metadata = dirfd (dir_metadata);
+
+  mkdirat (dfd_metadata, "gnu", S_IRWXU);
+
+  std::string path_gnu = path_metadata + "/gnu";
+  DIR *dir_gnu = opendir (path_gnu.c_str ());
+  if (dir_gnu == NULL)
+    {
+      closedir (dir_metadata);
+      return false;
+    }
+
+  int dfd_gnu = dirfd (dir_gnu);
+
+  std::string path_sha = "";
+  DIR *dir_sha = NULL;
+  if (hash_func_type == 0)
+    {
+      mkdirat (dfd_gnu, "gitoid_blob_sha1", S_IRWXU);
+
+      path_sha = path_gnu + "/gitoid_blob_sha1";
+      dir_sha = opendir (path_sha.c_str ());
+      if (dir_sha == NULL)
+	{
+	  closedir (dir_gnu);
+	  closedir (dir_metadata);
+	  return false;
+	}
+    }
+  else
+    {
+      mkdirat (dfd_gnu, "gitoid_blob_sha256", S_IRWXU);
+
+      path_sha = path_gnu + "/gitoid_blob_sha256";
+      dir_sha = opendir (path_sha.c_str ());
+      if (dir_sha == NULL)
+	{
+	  closedir (dir_gnu);
+	  closedir (dir_metadata);
+	  return false;
+	}
+    }
+
+  int is_E_or_S_or_c = 0;
+
+  std::string outfile_name =
+		omnibor_get_outfile_name (getenv ("COLLECT_GCC_OPTIONS"),
+						  &is_E_or_S_or_c);
+
+  if (outfile_name == "")
+    {
+      /* Option -o is not specified, so the name of the output file has
+	 to be deducted from the input file or be a.out in the case when
+	 linking is done as well.  */
+      if (pfile->deps->deps.size () > 0)
+	{
+	  /* Case when linking is done as well.  */
+	  if (is_E_or_S_or_c == 0)
+	    outfile_name = "a.out";
+	  /* Case when -c option is used.  */
+	  else if (is_E_or_S_or_c == 1)
+	    {
+	      std::string infile_name = pfile->deps->deps[0];
+	      /* TODO: Apart from supporting input file extensions with one
+		 character, support also '.cpp' extension.  */
+	      size_t start = infile_name.find_last_of ('/') + 1;
+	      outfile_name = infile_name.substr (start,
+						 infile_name.length() - 2 - start);
+	      outfile_name = outfile_name + ".o";
+	    }
+	  /* Case when -S option is used.  */
+	  else if (is_E_or_S_or_c == 2)
+	    {
+	      std::string infile_name = pfile->deps->deps[0];
+	      /* TODO: Apart from supporting input file extensions with one
+		 character, support also '.cpp' extension.  */
+	      size_t start = infile_name.find_last_of ('/') + 1;
+	      outfile_name = infile_name.substr (start,
+						 infile_name.length() - 2 - start);
+	      outfile_name = outfile_name + ".s";
+	    }
+	  /* Case when -E option is used.  In this case, the output file
+	     does not exist, because the preprocessed file will be
+	     outputed to stdout or stderr.  */
+	  else
+	    outfile_name = "not_available";
+	}
+      else
+	outfile_name = "not_available";
+    }
+
```

Comment:
If outfile_name is "not_available", don't create the metadata files at all because the output file will not exist so there is no point for its metadata to exist either.